### PR TITLE
[msal-browser][msal-common] Instance Aware Authentication (updated)

### PR DESCRIPTION
**Pull request by pkanher617**
This PR implements the instance_aware feature for msal-browser. Specifically, it adds the authority object to the response and handles all parameters which come back in the fragment.

It also adds TokenResponse to the browser exports (as per PR #1551).

Replaces PR #1584.

### CODE REVIEW

File: lib/msal-browser/src/app/PublicClientApplication.ts [hunk 1]
```diff
@@ -576,8 +572,10 @@ export class PublicClientApplication {
      * @param authorityUri 
      */
     private async getClientConfiguration(authorityUri?: string): Promise<ClientConfiguration> {
-        const discoveredAuthority = authorityUri ? await AuthorityFactory.createDiscoveredInstance(authorityUri, this.config.system.networkClient) 
-            : await this.defaultAuthorityPromise;
+        const defaultAuthority = await this.defaultAuthorityPromise;
+        const shouldCreateAuthority = !!authorityUri && authorityUri !== defaultAuthority.canonicalAuthority;
```
Comment (tnorling) [inline lib/msal-browser/src/app/PublicClientApplication.ts:576]: Are `authorityUri` and `defaultAuthority.canonicalAuthority` normalized somewhere? For example: Ensuring trailing slashes are consistent? Both are lower cased? etc.

File: lib/msal-browser/src/interaction_handler/InteractionHandler.ts [hunk 2]
```diff
@@ -43,13 +43,23 @@ export abstract class InteractionHandler {
...
         // Handle code response.
-        const authCode = this.authModule.handleFragmentResponse(locationHash, requestState);
+        const authCodeResponse: AuthorizationCodePayload = this.authModule.handleFragmentResponse(locationHash, requestState);
 
         // Assign code to request
-        this.authCodeRequest.code = authCode;
+        this.authCodeRequest.code = authCodeResponse.code;
+        if (authCodeResponse.cloud_graph_host_name) {
+            const cloudInstanceAuthorityUri = `https://${authCodeResponse.cloud_instance_host_name}/common/`;
+            if (cloudInstanceAuthorityUri !== this.browserStorage.getCachedAuthority()) {
```
Comment (tnorling) [inline lib/msal-browser/src/interaction_handler/InteractionHandler.ts:52]: Will the cached authority be cached with common or the specific tenant? Also same question here about whether the 2 things being compared here are normalized?

File: lib/msal-browser/src/interaction_handler/RedirectHandler.ts [hunk 3]
```diff
@@ -62,14 +62,26 @@ export class RedirectHandler extends InteractionHandler {
...
-        this.authCodeRequest.code = authCode;
+        const authCodeResponse: AuthorizationCodePayload = this.authModule.handleFragmentResponse(locationHash, requestState);
+
+        // Assign code to request
+        this.authCodeRequest.code = authCodeResponse.code;
+        if (authCodeResponse.cloud_graph_host_name) {
+            const cloudInstanceAuthorityUri = `https://${authCodeResponse.cloud_instance_host_name}/common/`;
+            if (cloudInstanceAuthorityUri !== this.browserStorage.getCachedAuthority()) {
+                const cloudInstanceAuthority = await AuthorityFactory.createDiscoveredInstance(this.authCodeRequest.authority, networkModule, true);
+                this.authModule.updateAuthority(cloudInstanceAuthority);
```
[the remark this answers is not in the document]
Comment (tnorling) [reply lib/msal-browser/src/interaction_handler/RedirectHandler.ts:73]: +1

File: lib/msal-common/src/response/AuthenticationResult.ts [hunk 4]
```diff
@@ -22,4 +23,6 @@ export class AuthenticationResult {
     extExpiresOn?: Date;
     state?: string;
     familyId?: string;
-}
+    cloudGraphHostName?: string;
+    msGraphHost?: string;
```
Comment (tnorling) [inline lib/msal-common/src/response/AuthenticationResult.ts:27]: Difference between `cloudGraphHostName` and `msGraphHost` ?

File: samples/VanillaJSTestApp2.0/app/instance_aware/graph.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        .catch(error => console.log(error));
+}
+
+async function seeProfile() {
+    const currentAcc = myMSALObj.getAccountByUsername(username);
+    if (currentAcc) {
+        const response = await getTokenPopup(loginRequest, currentAcc).catch(error => {
+            console.log(error);
+        });
+        callMSGraph(graphConfig.graphMeEndpoint, response.accessToken, updateUI);
```
Comment (tnorling) [inline samples/VanillaJSTestApp2.0/app/instance_aware/graph.js:28]: Since we're returning the graph endpoint in the token response should we demonstrate using that here?

File: lib/msal-browser/src/app/PublicClientApplication.ts (hunk 1 above)
Comment (pkanher617) [reply lib/msal-browser/src/app/PublicClientApplication.ts:576]: `defaultAuthority.canonicalAuthority` will always have a trailing slash, we should validate/normalize the given authorityUri however.

File: lib/msal-browser/src/interaction_handler/InteractionHandler.ts (hunk 2 above)
Comment (pkanher617) [reply lib/msal-browser/src/interaction_handler/InteractionHandler.ts:52]: I believe it is common but I'm not clear on this, need to consult docs for this.

File: lib/msal-common/src/client/AuthorizationCodeClient.ts [hunk 6]
```diff
@@ -79,16 +80,21 @@ export class AuthorizationCodeClient extends BaseClient {
...
         const hashUrlString = new UrlString(hashFragment);
         const serverParams = hashUrlString.getDeserializedHash<ServerAuthorizationCodeResponse>();
 
         // Get code response
         responseHandler.validateServerAuthorizationCodeResponse(serverParams, cachedState, this.cryptoUtils);
-        return serverParams.code;
+
+        return {
+            ...serverParams,
+            code: serverParams.code
```
[the remark this answers is not in the document]
Comment (pkanher617) [reply lib/msal-common/src/client/AuthorizationCodeClient.ts:96]: `code` is optional in ServerParams and required in `AuthorizationCodePayload`, which is why I have to explicitly set it here.

File: lib/msal-common/src/response/AuthenticationResult.ts (hunk 4 above)
Comment (pkanher617) [reply lib/msal-common/src/response/AuthenticationResult.ts:27]: not entirely sure, both come back as separate params though so I forwarded them.

File: lib/msal-common/src/client/AuthorizationCodeClient.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (jasonnutter) [reply lib/msal-common/src/client/AuthorizationCodeClient.ts:96]: Ah, gotya. Can you add that as a comment please?

**Comment by jasonnutter** [comment]
@pkanher617 Status on this PR?

**Comment by pkanher617** [comment]
> @pkanher617 Status on this PR?

Need to update this to latest dev and follow up with proper caching schema. Will update in the next week.

File: lib/msal-browser/src/interaction_handler/RedirectHandler.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (pkanher617) [reply lib/msal-browser/src/interaction_handler/RedirectHandler.ts]: done. reverted some changes that were previously made to this class though.

File: lib/msal-common/src/client/AuthorizationCodeClient.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (pkanher617) [reply lib/msal-common/src/client/AuthorizationCodeClient.ts:99]: done.

File: lib/msal-browser/src/app/PublicClientApplication.ts (hunk 1 above)
Comment (pkanher617) [reply lib/msal-browser/src/app/PublicClientApplication.ts]: verified that the authorities will always be consistent format and have trailing slashes.

File: lib/msal-browser/src/interaction_handler/InteractionHandler.ts (hunk 2 above)
Comment (pkanher617) [reply lib/msal-browser/src/interaction_handler/InteractionHandler.ts]: I have updated this to be cached with the tenant-id of the tenant used in the authorize call.

File: lib/msal-common/src/response/AuthenticationResult.ts (hunk 4 above)
Comment (pkanher617) [reply lib/msal-common/src/response/AuthenticationResult.ts:39]: cloudGraphHostname = AAD Graph, msGraphHost = MS Graph

File: samples/VanillaJSTestApp2.0/app/instance_aware/graph.js (hunk 5 above)
Comment (pkanher617) [reply samples/VanillaJSTestApp2.0/app/instance_aware/graph.js]: Done.

File: lib/msal-browser/src/app/ClientApplication.ts [hunk 7]
```diff
@@ -855,7 +855,7 @@ export abstract class ClientApplication {
             ...this.initializeBaseRequest(validatedRequest)
         };
 
-        this.browserStorage.updateCacheEntries(validatedRequest.state, validatedRequest.nonce, validatedRequest.authority);
+        this.browserStorage.updateCacheEntries(validatedRequest.state, validatedRequest.nonce, validatedRequest.authority || this.defaultAuthority.canonicalAuthority);
```
Comment (tnorling) [inline lib/msal-browser/src/app/ClientApplication.ts:858]: `validatedRequest` should already set a default authority, why is this needed?

File: lib/msal-common/src/response/AuthorizationCodePayload.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+/**
+ * Response returned after processing the code response query string or fragment.
+ */
+export type AuthorizationCodePayload = {
+    code: string,
+    cloud_instance_name?: string;
+    cloud_instance_host_name?: string;
+    cloud_graph_host_name?: string;
+    msgraph_host?: string;
+    state?: string;
```
Comment (tnorling) [inline lib/msal-common/src/response/AuthorizationCodePayload.ts:15]: state and nonce should be required, no?

File: samples/msal-browser-samples/VanillaJSTestApp2.0/app/instanceAware/authConfig.js [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+            }
+        }
+    }
+};
+
+// Add here scopes for id token to be used at MS Identity Platform endpoints.
+const loginRequest = {
+    scopes: ["User.Read"],
+    extraQueryParameters: {
+        "instance_aware": "true"
```
Comment (tnorling) [inline samples/msal-browser-samples/VanillaJSTestApp2.0/app/instanceAware/authConfig.js:40]: Does this need its own sample? Can we write a short doc instead?

**Review by tnorling — CHANGES_REQUESTED** [review]
Generally implementation looks ok, a few comments and would like to see unit tests. Also curious if the msal-common additions are type safe, given stricter type checking is imminent.

File: lib/msal-browser/src/app/ClientApplication.ts (hunk 7 above)
Comment (pkanher617) [reply lib/msal-browser/src/app/ClientApplication.ts:858]: good catch. I've removed it.

File: samples/msal-browser-samples/VanillaJSTestApp2.0/app/instanceAware/authConfig.js (hunk 9 above)
Comment (pkanher617) [reply samples/msal-browser-samples/VanillaJSTestApp2.0/app/instanceAware/authConfig.js:40]: Would prefer to have a sample that shows how to use the flow and the graph endpoints correctly. Happy to revisit if we have a lot of pushback on this.

Comment (tnorling) [reply samples/msal-browser-samples/VanillaJSTestApp2.0/app/instanceAware/authConfig.js:40]: I think we should write a doc (or update an existing one) for this regardless. The sample may or may not become redundant at that point

**Comment by pkanher617** [comment]
> Generally implementation looks ok, a few comments and would like to see unit tests. 
Unit tests are updated, and I am adding more for the scenarios around instance_aware. Also going to add e2e tests if I can.

> Also curious if the msal-common additions are type safe, given stricter type checking is imminent.
Would like to merge the strictNullChecks PR before I fully confirm this, but I think we should be fine.

File: samples/msal-browser-samples/VanillaJSTestApp2.0/app/instanceAware/authConfig.js (hunk 9 above)
Comment (pkanher617) [reply samples/msal-browser-samples/VanillaJSTestApp2.0/app/instanceAware/authConfig.js:40]: Will definitely update the docs to add the instance_aware flow

File: lib/msal-browser/src/interaction_handler/InteractionHandler.ts [hunk 10]
```diff
@@ -44,18 +44,34 @@ export abstract class InteractionHandler {
...
-        const authCode = this.authModule.handleFragmentResponse(locationHash, requestState);
+        const authCodeResponse = this.authModule.handleFragmentResponse(locationHash, requestState);
 
         // Get cached items
         const nonceKey = this.browserStorage.generateNonceKey(requestState);
         const cachedNonce = this.browserStorage.getTemporaryCache(nonceKey);
 
         // Assign code to request
-        this.authCodeRequest.code = authCode;
+        this.authCodeRequest.code = authCodeResponse.code;
```
[the remark this answers is not in the document]
Comment (pkanher617) [reply lib/msal-browser/src/interaction_handler/InteractionHandler.ts:54]: Yes that's correct. If you have suggestions please suggest, but I think it is clear enough. The authCodeRequest is a field and the response is not.

File: lib/msal-browser/src/interaction_handler/RedirectHandler.ts [hunk 11]
```diff
@@ -3,31 +3,28 @@
...
+        super(authCodeModule, storageImpl);
         this.browserCrypto = browserCrypto;
     }
 
     /**
      * Redirects window to given URL.
      * @param urlNavigate
      */
-    initiateAuthRequest(requestUrl: string, authCodeRequest: AuthorizationCodeRequest, redirectTimeout: number, redirectStartPage?: string): Promise<void> {
+    initiateAuthRequest(requestUrl: string, authCodeRequest: AuthorizationCodeRequest, redirectTimeout?: number, redirectStartPage?: string): Promise<void> {
```
Comment (jasonnutter) [inline lib/msal-browser/src/interaction_handler/RedirectHandler.ts:27]: Why is `redirectTimeout` set to optional? FYI in #2669, I also made `redirectStartPage` required, since it will always be passed: https://github.com/AzureAD/microsoft-authentication-library-for-js/pull/2669/files#diff-d592fb8a2a52b3d36d8625f2210c27b8360c6d1f0cff4fe06494d1baf77ac593

Comment (jasonnutter) [reply lib/msal-browser/src/interaction_handler/RedirectHandler.ts:27]: Oh, is it because it now implements `InteractionHandler`?

**Review by jasonnutter — APPROVED** [review]
One question otherwise looks good!

File: lib/msal-browser/src/interaction_handler/RedirectHandler.ts (hunk 11 above)
Comment (pkanher617) [reply lib/msal-browser/src/interaction_handler/RedirectHandler.ts:27]: Yeah that's correct. The interface implementation does not include redirectTimeout.

File: lib/msal-common/src/response/AuthorizationCodePayload.ts (hunk 8 above)
Comment (pkanher617) [reply lib/msal-common/src/response/AuthorizationCodePayload.ts:15]: No because these are set by the platform library. They are not required in msal-node.

File: lib/msal-browser/src/interaction_handler/RedirectHandler.ts (hunk 11 above)
Comment (tnorling) [reply lib/msal-browser/src/interaction_handler/RedirectHandler.ts:27]: Can we remove the `initiateAuthRequest` abstract from `InteractionHandler` so that we dont have to make these optional?

File: lib/msal-browser/src/interaction_handler/InteractionHandler.ts [hunk 12]
```diff
@@ -44,18 +44,34 @@ export abstract class InteractionHandler {
...
         // Acquire token with retrieved code.
-        const tokenResponse = await this.authModule.acquireToken(this.authCodeRequest, cachedNonce, requestState);
+        const tokenResponse = await this.authModule.acquireToken(this.authCodeRequest, authCodeResponse);
         this.browserStorage.cleanRequest(serverParams.state);
         return tokenResponse;
     }
+
+    protected async updateTokenEndpointAuthority(cloudInstanceHostname: string, authority: Authority, networkModule: INetworkModule): Promise<void> {
+        const cloudInstanceAuthorityUri = `https://${cloudInstanceHostname}/${authority.tenant}/`;
+        if (cloudInstanceAuthorityUri !== authority.canonicalAuthority) {
```
[the remark this answers is not in the document]
Comment (pkanher617) [reply lib/msal-browser/src/interaction_handler/InteractionHandler.ts]: done.

File: lib/msal-browser/src/interaction_handler/RedirectHandler.ts (hunk 11 above)
Comment (pkanher617) [reply lib/msal-browser/src/interaction_handler/RedirectHandler.ts:27]: I would like to keep the abstract function as a constraint for subclasses implementing this. I understand that there needs to be a compiler level way to ensure that the redirectTimeout and redirectStartPage are passed, but for now can we ensure this by adding code errors?

Comment (tnorling) [reply lib/msal-browser/src/interaction_handler/RedirectHandler.ts:27]: I personally don't like making a param optional but then throwing if it's not provided. Would prefer either making this its own function that doesn't implement the abstract or getting rid of the abstract altogether.

Comment (pkanher617) [reply lib/msal-browser/src/interaction_handler/RedirectHandler.ts:27]: I personally think it's worse design to take out the abstract functionality than to check if an optional parameter is provided or not. Let me see if I can find an alternative, but in the meantime can you approve the PR so I can merge as soon as it is ready?

Comment (jasonnutter) [reply lib/msal-browser/src/interaction_handler/RedirectHandler.ts:27]: I also don't like the idea of throwing an error if the params are not passed. I would rather the code just handle it gracefully.

Comment (jasonnutter) [reply lib/msal-browser/src/interaction_handler/RedirectHandler.ts:27]: So you can also just leave this as you have it, and I can deal with adjustments that need to be made for my PR.

Comment (pkanher617) [reply lib/msal-browser/src/interaction_handler/RedirectHandler.ts]: made an update to the `initiateAuthRequest` function, please check and see if it works for you guys.

File: lib/msal-browser/src/app/ClientApplication.ts [hunk 13]
```diff
@@ -303,7 +303,10 @@ export abstract class ClientApplication {
 
             const redirectStartPage = (request && request.redirectStartPage) || window.location.href;
             // Show the UI once the url has been created. Response will come back in the hash, which will be handled in the handleRedirectCallback function.
-            return interactionHandler.initiateAuthRequest(navigateUrl, authCodeRequest, this.config.system.redirectNavigationTimeout, redirectStartPage);
+            return interactionHandler.initiateAuthRequest(navigateUrl, authCodeRequest, {
```
Comment (tnorling) [inline lib/msal-browser/src/app/ClientApplication.ts:306]: nit: Can we define the params above and make this a one liner?

File: lib/msal-browser/src/app/ClientApplication.ts [hunk 14]
```diff
@@ -379,7 +382,9 @@ export abstract class ClientApplication {
             const interactionHandler = new PopupHandler(authClient, this.browserStorage);
 
             // Show the UI once the url has been created. Get the window handle for the popup.
-            const popupWindow: Window = interactionHandler.initiateAuthRequest(navigateUrl, authCodeRequest, popup);
+            const popupWindow: Window = interactionHandler.initiateAuthRequest(navigateUrl, authCodeRequest, {
```
Comment (tnorling) [inline lib/msal-browser/src/app/ClientApplication.ts:385]: Same here

File: lib/msal-common/src/response/AuthorizationCodePayload.ts (hunk 8 above)
Comment (tnorling) [reply lib/msal-common/src/response/AuthorizationCodePayload.ts:15]: Ah right, @sameerag  was going to add them to node and will change this then

**Review by tnorling — APPROVED** [review]
Looks good, thanks for addressing feedback

File: lib/msal-browser/src/app/ClientApplication.ts (hunk 13 above)
Comment (pkanher617) [reply lib/msal-browser/src/app/ClientApplication.ts]: done.

File: lib/msal-browser/src/app/ClientApplication.ts (hunk 14 above)
Comment (pkanher617) [reply lib/msal-browser/src/app/ClientApplication.ts]: done.

**Review by jmckennon — APPROVED** [review]
LGTM!
